### PR TITLE
Mocks return arrays

### PR DESCRIPTION
**Pull request by neuromantic99**

### CODE REVIEW

File: lib/prompts/apiMock.ts [hunk 1]
```diff
@@ -76,9 +90,19 @@ CEO name (string): The name of the CEO
...
+CEO name: [Tim Cook, Sundar Pichai, Mark Zuckerberg]
+`
+    : `
+Company name: Apple
 Annual earnings: 1000000
 CEO name: Tim Cook
+`
+}
 
 -- END OF EXAMPLES --
```
Comment (Henry-Pulver) [inline lib/prompts/apiMock.ts:108]: Please could we write some tests for the line below?

I'm not really sure how it works for objects :confused: 

E.g. if there's a `user` object, would it output `User (object): <description>` as one line? 

```
${Object.entries(responseType)
  .map(([k, v]) => `${k} (${v.type}): ${v.description ?? "no description"}`)
  .join("\n")}
```

File: lib/utils.ts [hunk 2]
```diff
@@ -288,7 +291,21 @@ export function transformProperties(chunks: Chunk[]): Properties {
     const existingProperty = properties[fieldName] ?? { path: chunk.path };
 
     if (["type", "description"].includes(chunkType?.toString() ?? "")) {
-      properties[fieldName] = { ...existingProperty, [chunkType]: chunk.data };
+      // If statement below is true when the chunk is an array of primitives. Without this check,
+      // we ask gpt for a value for the key "item" which is part of the schema
+      // structure, not data we want returned.
+      // TODO: this doesn't treat the array as an array currently
+      if (fieldName === "items" && !["object", "array"].includes(chunk.data)) {
```
Comment (Henry-Pulver) [inline lib/utils.ts:298]: This boggles my mind :confused: 

Can keep for now but the number of steps involved in this conversion makes it tricky to follow

Comment above is good though :muscle:

File: lib/utils.ts [hunk 3]
```diff
@@ -301,17 +318,40 @@ export function chunkToString(chunk: Chunk): string {
...
   properties: Properties,
-  gptOutput: string
+  gptOutput: string,
+  arrayIdx: number | null = null
 ): Properties {
   /**
   Add the data outputted by gpt to the properties object. 
+
+  If working with an array type (and gpt has output a comma separated list of values)
+  pass arrayIdx to select which value in an semi-colon separate  list to use.
```
Comment (Henry-Pulver) [inline lib/utils.ts:328]: Love these new comments :smile:

File: lib/utils.ts [hunk 4]
```diff
@@ -301,17 +318,40 @@ export function chunkToString(chunk: Chunk): string {
...
+        value = value
+          .replace("[", "")
+          .replaceAll("]", "")
+          .split(",")
+          [arrayIdx].trim(); // TODO: Add fallback for if the commaIdx is out of range
+      }
+
+      if (!value) return;
+
+      value = value.replace(/^["'](.+(?=["']$))["']$/, "$1");
```
Comment (Henry-Pulver) [inline lib/utils.ts:350]: Short comment?

File: pages/api/mock/[...slug].ts [hunk 5]
```diff
@@ -207,17 +208,19 @@ export default async function handler(
         ]
       : null;
 
-  // TODO: properties can be set as array or object by the schema, currently not
-  // dealing with this
   const properties = schema ? propertiesFromSchema(schema) : null;
 
+  // Just deal with the first layer of nesting for now
+  const isArray = schema?.type === "array";
```
Comment (Henry-Pulver) [inline pages/api/mock/[...slug].ts:214]: What actually happens if we are mocking an array of arrays?

File: pages/api/mock/[...slug].ts [hunk 6]
```diff
@@ -253,26 +256,40 @@ export async function getMockedProperties(
...
+        : Math.random() >= 0.5;
+      hardCodedProperties[key] = {
+        data: value.data,
+        type: "boolean",
+        path: value.path,
+      };
+    } else {
+      propertiesForAi[key] = value;
+    }
+  }
```
Comment (Henry-Pulver) [inline pages/api/mock/[...slug].ts:282]: Cheeky comment saying why some are hardcoded and others not might be useful :+1:

File: pages/api/mock/[...slug].ts [hunk 7]
```diff
@@ -286,13 +303,71 @@ export async function getMockedProperties(
...
+      };
+    }
+
+    const newChunks = propertiesToChunks(
+      Object.assign({}, transformed, { ...readded, ...hardcodedProperiesIdx })
+    );
+    arrayResponse.push(jsonReconstruct(newChunks));
+  }
+  return arrayResponse;
+}
```
Comment (Henry-Pulver) [inline pages/api/mock/[...slug].ts:369]: I'll be honest. In this file, there are so many data formats and functions manipulating between these formats that this function loses me.

Tests would make me feel 2x better about this :pray:

**Review by Henry-Pulver — COMMENTED** [review]
You did tell me this would be a tricky one and you were certainly right :sweat_smile:

File: lib/prompts/apiMock.ts (hunk 1 above)
Comment (neuromantic99) [reply lib/prompts/apiMock.ts:108]: Objects are already deconstrucuted to a "Properties" type, so this line doesn't need to deal with nested objects, only flat "properties". I've added a test in apiMockPrompt.test.ts to demonstrate this.

File: pages/api/mock/[...slug].ts (hunk 5 above)
Comment (neuromantic99) [reply pages/api/mock/[...slug].ts:214]: updated the comment

File: __tests__/apiMock/chunksToProperties.test.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  it("merges chunks with the same path", () => {
+    const chunks: Chunk[] = [
+      { path: ["id", "type"], data: "integer" },
+      { path: ["id", "description"], data: "Customer id" },
+      { path: ["id", "format"], data: "int64" },
+    ];
+
+    const properties: Properties = chunksToProperties(chunks);
+    expect(properties).toEqual({
+      id: { type: "integer", description: "Customer id", path: ["id"] },
```
Comment (Henry-Pulver) [inline __tests__/apiMock/chunksToProperties.test.ts:60]: Confused by this test. Why does `format` disappear?

File: __tests__/apiMock/getMockedProperties.test.ts [hunk 9]
```diff
@@ -43,66 +43,55 @@ lastName: Smith
...
   it("nested object", async () => {
+    // TODO: remove properties from here
+    // learney.atlassian.net/browse/SF-2007
     const openAiResponse = `
   browserSdkVersion: 1
   dateCreated: 2021-01-01
-  cdn: nice-cdn
-  csp: nice-csp
+  dsn.properties.cdn: nice-cdn
+  dsn.properties.csp: nice-csp
```
Comment (Henry-Pulver) [inline __tests__/apiMock/getMockedProperties.test.ts:94]: So this still doesn't remove the `properties` and `items` keywords from the generated json? Or am I missing a step where this happens?

File: pages/api/mock/[...slug].ts [hunk 10]
```diff
@@ -246,33 +250,51 @@ export default async function handler(
...
-    .filter(
-      ([_, value]) =>
-        value.type?.toLowerCase() !== "object" &&
-        value.type?.toLowerCase() !== "array"
-    )
-    .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
+  const chunks = jsonSplitter(openApiProperties);
+  const allProperties = chunksToProperties(chunks);
+
+  console.log("allProperties", allProperties);
```
Comment (Henry-Pulver) [inline pages/api/mock/[...slug].ts:266]: This deliberate? Fine if it is, just wasn't sure :sweat_smile:

**Comment by Henry-Pulver** [comment]
Okay, I'll come clean - I didn't actually test it out on staging... until now :astonished: 

Don't think this is expected behaviour here
Screenshot from 2023-08-09 21-15-13

File: __tests__/apiMock/getMockedProperties.test.ts (hunk 9 above)
Comment (neuromantic99) [reply __tests__/apiMock/getMockedProperties.test.ts:94]: properties is still there, there's a ticket for this

https://learney.atlassian.net/jira/software/projects/SF/boards/2/backlog?assignee=627e349e98eae500689fcf62&selectedIssue=SF-2007

File: __tests__/apiMock/chunksToProperties.test.ts (hunk 8 above)
Comment (neuromantic99) [reply __tests__/apiMock/chunksToProperties.test.ts:60]: we're not currently taking format into account when generating the response, only type

Comment (Henry-Pulver) [reply __tests__/apiMock/chunksToProperties.test.ts:60]: Oh ok I thought `Properties` was a general way of reformulating json that would works on all jsons - so it's actually OpenAPI specific?

Comment (neuromantic99) [reply __tests__/apiMock/chunksToProperties.test.ts:60]: yes queen, jsonDeconstruct / Reconstruct is general

**Pull request merged**